src/app/InteractionModelDelegate.h

```diff
@@ -0,0 +1,118 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Interaction Model Delegate.
+ *
+ */
+
+#pragma once
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+
+struct EventPathParams;
+
+/**
+ * @brief
+ *   This class provides a skeleton for the callback functions. The functions will be
+ *   called by InteractionModel on specific CallbackId. If the user of InteractionModel
+ *   is interested in receiving these callbacks, they can specialize this class and handle
+ *   each CallbackId in their implementation of this class.
+ */
+class InteractionModelDelegate
+{
+public:
+    virtual ~InteractionModelDelegate() {}
+
+    // Interaction Model would use the below callbackID as discriminator and notify the zcl/app the particular interaction
+    // has completed or is happening or need more things from upper layer via InParam and OutParam
+    // The below callbacks may or may not be necessary for upper zcl, once the zcl integration is complete, we will revisit these
+    enum class CallbackId
+    {
+        // Read Client
+        /**
+         * Sent when it is ready to send the ReadRequest.  The
+         * application is expected to fill in the attribute/event paths that it wants to
+         * read/subscribe to.
+         */
+        kReadRequestPrepareNeeded,
+        // IM notify zcl  events can be further processed in app.
+        kEventStreamReceived,
+
+        // IM notify zcl all reports have been processed.
+        kReportProcessed,
+    };
+
+    /**
+     * Incoming parameters sent with callback generated directly from this component
+     *
+     */
+    union InParam
```

Comment:
I was having the same question, especially since the union currently only has one member.

What is the long-term expected API shape here?  Are all possible types of OutParam valid for all possible types of InParam?  Or are some specific types involved, depending on the callbackid value?

Looking around it seems like the basic structure of a readrequest is that someone tells a ReadClient "Send the Read Request", and the ReadClient then turns around and asks the `InteractionModelDelegate` "what are we reading, exactly?"

But why is that the setup instead of just having SendReadRequest take the things we want to read as an argument in some form?  Or if we absolutely must fill them in programmatically somehow take a callback to do that, instead of the ReadClient making a virtual callback that then has to examine the passed-in enum to see what it's being asked and mess around with the in/out unions....

I'd really like to understand why this setup is the way it is.

Reply:
previsouly i am thinking of the vtable size, currently we have defined 11 messages in IM encoding spec, assume for each message's implementation(client/handler), we need at least one callbacks to hook to zcl/app,. For example, for invoke command request,  we also need to notify zcl all commands have been processed after processing all incoming command response, for read request, we need to ask for path info from zcl app, and notify all reports have processed. For subscribe request, we also need to ask for path info from zcl app, notify zcl that subscription is established since subscribe response is sent after all Reports have been sent back to the client.  Fo read and command and write handler/receiver, it all needs to do acl check for particular command path/attribute path /event path. for write handler, when all data has been loaded, and it may notify zcl to propagate the change to necessary UI or persistent store.

regarding HandlePathPrepareNeededCallBack, current read and subscribe client would use this callback. In resubscribe, internally timer may need to resubscribe when previous subscription goes away, there send request may not take path info as parameter, we need this callback to retrieve path info. I will create one ticket to continue this discussion after subscribe is implemented.

Reply:
after chatting with @bzbarsky-apple, we decided to use virtual fucntions instead of unions. 

In addition,  once we have more context and code development on subscription and later attribures for how the delegate's subclasses look we can figure out whether that's a sensible approach....maybe it does make sense to just have a ReadDelegate, SubscribeDelegate, etc. I will create the ticket for this effort